### PR TITLE
[Java] Avoid data copy from C++ to Java for ByteBuffer type

### PR DESCRIPTION
**Pull request by kfstorm**
## Why are these changes needed?

If the type of one or more parameters is `ByteBuffer`, we create a direct `ByteBuffer` instead in Java to avoid an extra data copy operation (compared to the old-fashioned way of creating a `byte[]` instance with `NewByteArray` in JNI code).

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: java/runtime/src/main/java/io/ray/runtime/task/TaskExecutor.java [hunk 1]
```diff
@@ -61,10 +64,34 @@ void setActorContext(T actorContext) {
...
+    try {
+      localRayFunction.set(getRayFunction(rayFunctionInfo));
+    } catch (Throwable e) {
+      // Ignore the exception.
+      return new boolean[0];
+    }
+    Class<?>[] types = localRayFunction.get().executable.getParameterTypes();
+    boolean[] results = new boolean[types.length];
+    for (int i = 0; i < types.length; i++) {
+      results[i] = types[i] == ByteBuffer.class;
```
[the remark this answers is not in the document]
Comment (kfstorm) [reply java/runtime/src/main/java/io/ray/runtime/task/TaskExecutor.java:87]: `DirectByteBuffer` is not a class. Instead, `ByteBuffer` has a method called `isDirect`.

File: java/runtime/src/main/java/io/ray/runtime/task/TaskExecutor.java [hunk 2]
```diff
@@ -61,10 +64,34 @@ void setActorContext(T actorContext) {
...
+  private RayFunction getRayFunction(List<String> rayFunctionInfo) {
+    JobId jobId = runtime.getWorkerContext().getCurrentJobId();
+    JavaFunctionDescriptor functionDescriptor = parseFunctionDescriptor(rayFunctionInfo);
+    return runtime.getFunctionManager().getFunction(jobId, functionDescriptor);
+  }
+
+  /**
+   * The return value indicates which parameters are ByteBuffer.
+   */
+  protected boolean[] parseFunctionArguments(List<String> rayFunctionInfo) {
```
[the remark this answers is not in the document]
Comment (kfstorm) [reply java/runtime/src/main/java/io/ray/runtime/task/TaskExecutor.java]: Done.

File: src/ray/core_worker/lib/java/io_ray_runtime_RayNativeRuntime.cc [hunk 3]
```diff
@@ -78,8 +78,37 @@ JNIEXPORT void JNICALL Java_io_ray_runtime_RayNativeRuntime_nativeInitialize(
...
+        jbooleanArray java_pre_execute_results =
+            static_cast<jbooleanArray>(env->CallObjectMethod(
+                java_task_executor, java_task_executor_parse_function_arguments,
+                ray_function_array_list));
+        RAY_CHECK_JAVA_EXCEPTION(env);
+        jboolean *pre_execute_results =
+            env->GetBooleanArrayElements(java_pre_execute_results, nullptr);
+        jobject args_array_list = nullptr;
+        // If the lengths don't match, it means that `parseFunctionArguments` failed. In
+        // this case, pass null as arguments to `execute` method.
```
[the remark this answers is not in the document]
Comment (kfstorm) [reply src/ray/core_worker/lib/java/io_ray_runtime_RayNativeRuntime.cc]: Done. I've refined the comment.

File: src/ray/core_worker/lib/java/io_ray_runtime_RayNativeRuntime.cc [hunk 4]
```diff
@@ -78,8 +78,37 @@ JNIEXPORT void JNICALL Java_io_ray_runtime_RayNativeRuntime_nativeInitialize(
...
+                  jobject obj = env->NewDirectByteBuffer(
+                      native_object->GetData()->Data(), native_object->GetData()->Size());
+                  RAY_CHECK(obj);
+                  return obj;
+                }
+                return NativeRayObjectToJavaNativeRayObject(env, native_object);
+              });
+        }
+        env->ReleaseBooleanArrayElements(java_pre_execute_results, pre_execute_results,
+                                         JNI_ABORT);
```
[the remark this answers is not in the document]
Comment (kfstorm) [reply src/ray/core_worker/lib/java/io_ray_runtime_RayNativeRuntime.cc]: Done.

**Review by raulchen — APPROVED** [review]

**Pull request merged**